// bbot/dist/lib/logger.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const winston = __importStar(require("winston"));
const { combine, timestamp, json, colorize, align } = winston.format;
exports.logger = winston.createLogger({
    level: process.env.BOT_LOG_LEVEL,
    transports: [
        new winston.transports.File({
            filename: 'error.log',
            level: 'error',
            maxsize: 500000,
            format: combine(timestamp(), json())
        }),
        new winston.transports.File({
            filename: 'combined.log',
            level: 'debug',
            maxsize: 500000,
            format: combine(timestamp(), json())
        }),
        new winston.transports.Console({
            format: combine(colorize(), align(), winston.format.printf((nfo) => {
                return `${nfo.level}: ${nfo.message}`;
            }))
        })
    ],
    exceptionHandlers: [
        new winston.transports.File({ filename: 'exceptions.log' })
    ],
    exitOnError: (err) => (typeof err.middleware === 'undefined')
});
