// bbot/dist/lib/payload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Payload {
    constructor(meta = {}) {
        if (meta.attachments)
            for (let i of meta.attachments)
                this.attachment(i);
        if (meta.actions)
            for (let i of meta.actions)
                this.action(i);
        if (meta.quickReplies)
            for (let i of meta.quickReplies)
                this.quickReply(i);
    }
    custom(object) {
        for (let key in object)
            this[key] = object[key];
    }
    attachment(attachment) {
        if (!this.attachments)
            this.attachments = [];
        this.attachments.push(attachment);
        return this;
    }
    action(action) {
        if (!this.actions)
            this.actions = [];
        this.actions.push(action);
        return this;
    }
    quickReply(quickReply) {
        if (!quickReply.type)
            quickReply.type = 'button';
        if (!quickReply.content)
            quickReply.content = quickReply.text;
        if (!this.quickReplies)
            this.quickReplies = [];
        this.quickReplies.push(quickReply);
        return this;
    }
    toObject() {
        return JSON.parse(JSON.stringify(this));
    }
}
exports.Payload = Payload;
